src/pybind/mgr/cephadm/inventory.py

```diff
@@ -1699,6 +1764,294 @@ def agent_config_successfully_delivered(self, daemon_spec: CephadmDaemonDeploySp
+            'iscsi_ssl_key': {},  # service-name -> key
+            'ingress_ssl_key': {},  # service-name -> key
+        }
+
+    def get_cert(self, entity: str, service_name: str = '', host: str = '') -> str:
+        self._validate_cert_entity(entity, service_name, host)
+
+        cert = Cert()
+        if entity in self.service_name_cert or entity in self.host_cert:
+            var = service_name if entity in self.service_name_cert else host
+            if var not in self.known_certs[entity]:
+                return ''
+            cert = self.known_certs[entity][var]
+        else:
+            cert = self.known_certs[entity]
+        if not cert or not isinstance(cert, Cert):
+            return ''
+        return cert.cert
+
+    def save_cert(self, entity: str, cert: str, service_name: str = '', host: str = '', user_made: bool = False) -> None:
+        self._validate_cert_entity(entity, service_name, host)
+
+        cert_obj = Cert(cert, user_made)
+
+        j: Union[str, Dict[Any, Any], None] = None
+        if entity in self.service_name_cert or entity in self.host_cert:
+            var = service_name if entity in self.service_name_cert else host
+            j = {}
+            self.known_certs[entity][var] = cert_obj
+            for service_name in self.known_certs[entity].keys():
+                j[var] = Cert.to_json(self.known_certs[entity][var])
+        else:
+            self.known_certs[entity] = cert_obj
+            j = Cert.to_json(cert_obj)
+        self.mgr.set_store(CERT_STORE_CERT_PREFIX + entity, json.dumps(j))
+
+    def rm_cert(self, entity: str, service_name: str = '', host: str = '') -> None:
+        self.save_cert(entity, cert='', service_name=service_name, host=host)
+
+    def _validate_cert_entity(self, entity: str, service_name: str = '', host: str = '') -> None:
+        if entity not in self.known_certs.keys():
+            raise OrchestratorError(f'Attempted to access cert for unknown entity {entity}')
+
+        if entity in self.host_cert and not host:
+            raise OrchestratorError(f'Need host to access cert for entity {entity}')
+
+        if entity in self.service_name_cert and not service_name:
+            raise OrchestratorError(f'Need service name to access cert for entity {entity}')
+
+    def cert_ls(self) -> Dict[str, Union[bool, Dict[str, bool]]]:
+        ls: Dict[str, Any] = {}
+        for k, v in self.known_certs.items():
+            if k in self.service_name_cert or k in self.host_cert:
+                tmp: Dict[str, Any] = {}
+                for key in self.known_certs[k]:
+                    if self.known_certs[k][key]:
+                        tmp[key] = True
+                if len(tmp) == 0:
+                    ls[k] = False
+                else:
+                    ls[k] = tmp
+            else:
+                if self.known_certs[k]:
+                    ls[k] = True
+                else:
+                    ls[k] = False
+        return ls
```

Comment:
Plz, can you explain with some code comment what are we trying to achieve here? I find the logic a little bit confusing.

Reply:
The outer loop goes through all the potential certs/keys we the cert store is aware of. `self.service_name_cert` is the list of these certs that is stored as a mapping from service names to certs. `self.host_cert` is the list of all certs that are stored as mappings from hosts to certs. Since those are both one level deep strings in the dict, you can handle the identically, hence the single loop going through the keys in `self.known_certs`. Beyond that, we are building this `ls` dict that basically just checks for each cert we could have if there is actually one there. Then `tmp` is just for tracking within a single type of cert (e.g. rgw_frontend_ssl_cert or ingress' ssl_cert) what we have for that cert. It then checks if that tmp dict actually is populated or not. If it's not (`if len(tmp) == 0`) then it just sets that key to false, and otherwise sets it to the tmp dict we just built. PArt of the reason it's so confusing is the structure it's trying to build is not intuitive. 
```
{
    "rgw_frontend_ssl_cert": false,
    "iscsi_ssl_cert": false,
    "nvmeof_client_cert": false,
    "nvmeof_server_cert": false,
    "ingress_ssl_cert": false,
    "agent_endpoint_root_cert": false,
    "service_discovery_root_cert": false,
    "grafana_cert": false,
    "alertmanager_cert": {
        "host1": true,
        "host2": true
    },
    "prometheus_cert": false,
    "node_exporter_cert": false
}
```
I will probably actually change this output later as discussed in https://github.com/ceph/ceph/pull/55564#discussion_r1643147887

Reply:
@adk3798 thank you very much for the detailed explanation. I think the code may be simplified to something as bellow (bool operator can help to evaluate dictionaries):

```
def cert_ls(self) -> Dict[str, Union[bool, Dict[str, bool]]]:
    ls = {}
    for k, v in self.known_certs.items():
        if k in self.service_name_cert or k in self.host_cert:
            tmp = {key: True for key in v if v[key]}
            ls[k] = tmp if tmp else False
        else:
            ls[k] = bool(v)
    return ls
```

Reply:
updated with this logic